### DISCUSSION TITLE
How to add custom validation on top of database validity

### DISCUSSION THREAD

**Discussion by splitbrain** [tflori/orm#85] [Q&A] [answered]
I've read https://tflori.github.io/orm/validate.html a couple of times now, but I am still not sure if the following is possible and if so how.

I want to keep the auto generated table description if possible, but want to add custom validation to some fields. Simplest use case: I have columns that do not allow NULL values, but an empty string is not NULL, so the validation does not complain. I would like to add a `trim($value) !== ''` check before a value is validated as okay.

Any hints on how to do this?

**Comment by tflori** [comment, marked as answer]
You are lucky I read everything twice. First I wrote a totally different answer that would just confuse more :smile: 

The easiest way to do so is to just write a own type:

```php
<?php

namespace App\Dbal\Type;

class NonEmptyVarChar extends ORM\Dbal\Type\VarChar
{
    public static function fits(array $columnDefinition)
    {
        return $columnDefinition['data_type'] === 'varchar'; // && substr($columnDefinition['column_name'], -2) === '_f'
    }

    public function validate($value)
    {
        $valid = parent::validate($value);
        if ($valid === true && trim($value) === '') {
            // has to be a ORM\Dbal\Error class....
            return new EmptyString('String should not be empty');
        }

        return $valid;
    }
}
```

Then you can register it with `ORM\Dbal\Column::registerType(NonEmptyVarChar::class)` and because it is prepended to the list of known types it will be used over the default one. If you still want to use the default validators for the other var char fields you might need to filter the columns where you want to use it within the `fits(array $columnDefinition)` method (example in the code).

Does that help?

**Comment by splitbrain** [comment]
Ah, the fits() mechanism is interesting. Yeah I guess this answers my question. Thank you.

PS: might be worth to enable github's discussion feature and move this issue there as it's not really a bug or feature request.

**Reply by tflori** [reply]
point taken, done.
